packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/FLTGoogleMapHeatmapController.m

```diff
@@ -0,0 +1,186 @@
+  [self setMap];
+}
+- (NSDictionary<NSString *, id> *)getHeatmapInfo {
+  NSMutableDictionary *options = [@{} mutableCopy];
+  options[kHeatmapDataKey] =
+      [FLTGoogleMapJSONConversions arrayFromWeightedData:_heatmapTileLayer.weightedData];
+  options[kHeatmapGradientKey] =
+      [FLTGoogleMapJSONConversions dictionaryFromGradient:_heatmapTileLayer.gradient];
+  options[kHeatmapOpacityKey] = @(_heatmapTileLayer.opacity);
+  options[kHeatmapRadiusKey] = @(_heatmapTileLayer.radius);
+  options[kHeatmapMinimumZoomIntensityKey] = @(_heatmapTileLayer.minimumZoomIntensity);
+  options[kHeatmapMaximumZoomIntensityKey] = @(_heatmapTileLayer.maximumZoomIntensity);
+  return options;
+}
+@end
+
+@interface FLTHeatmapsController ()
+
+// A map from heatmapId to the controller that manages it.
+@property(nonatomic, strong) NSMutableDictionary *heatmapIdToController;
+
+// The map view owned by GoogmeMapController.
+@property(nonatomic, weak) GMSMapView *mapView;
+
+@end
+
+@implementation FLTHeatmapsController
+- (instancetype)initWithMapView:(GMSMapView *)mapView {
+  self = [super init];
+  if (self) {
+    _mapView = mapView;
+    _heatmapIdToController = [@{} mutableCopy];
+  }
+  return self;
+}
+- (void)addHeatmaps:(NSArray<NSDictionary<NSString *, id> *> *)heatmapsToAdd {
+  for (NSDictionary<NSString *, id> *heatmap in heatmapsToAdd) {
+    NSString *heatmapId = [FLTHeatmapsController heatmapIdentifierFor:heatmap];
+    GMUHeatmapTileLayer *heatmapTileLayer = [[GMUHeatmapTileLayer alloc] init];
+    FLTGoogleMapHeatmapController *controller =
+        [[FLTGoogleMapHeatmapController alloc] initWithHeatmapTileLayer:heatmapTileLayer
+                                                                mapView:_mapView
+                                                                options:heatmap];
+    _heatmapIdToController[heatmapId] = controller;
+  }
+}
+- (void)changeHeatmaps:(NSArray<NSDictionary<NSString *, id> *> *)heatmapsToChange {
+  for (NSDictionary<NSString *, id> *heatmap in heatmapsToChange) {
+    NSString *heatmapId = [FLTHeatmapsController heatmapIdentifierFor:heatmap];
+    FLTGoogleMapHeatmapController *controller = _heatmapIdToController[heatmapId];
+
+    [controller interpretHeatmapOptions:heatmap];
+    [controller clearTileCache];
+  }
+}
+- (void)removeHeatmapsWithIdentifiers:(NSArray<NSString *> *)identifiers {
+  for (NSString *heatmapId in identifiers) {
+    FLTGoogleMapHeatmapController *controller = _heatmapIdToController[heatmapId];
+    if (!controller) {
+      continue;
+    }
+    [controller removeHeatmap];
+    [_heatmapIdToController removeObjectForKey:heatmapId];
+  }
+}
+- (bool)hasHeatmapWithIdentifier:(NSString *)identifier {
+  if (!identifier) {
+    return NO;
+  }
+  return _heatmapIdToController[identifier] != nil;
+}
+- (nullable NSDictionary<NSString *, id> *)heatmapInfoWithIdentifier:(NSString *)identifier {
+  if (self.heatmapIdToController[identifier] == nil) {
+    return nil;
+  }
```

Comment:
This check is not necessary; the line below will already return `nil` in this case.